### PR TITLE
e2e refactor

### LINKED ISSUE #66649: refactor test/e2e/framework
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

The E2E framework has grown organically over time. As a result, it is both hard to maintain and close to impossible to use outside of Kubernetes itself.

**What you expected to happen**:

The goal is to make it possible that other projects can:
- import test/e2e/framework as a vendor dependency without pulling in support for cloud providers and any tests (because of the huge dependency chains that a complete import causes)
- import support for cloud providers and tests separately if desired
- add custom tests without modifying the framework
- create and run a Ginkgo suite

One potential user for this is the Kubernetes CSI project, which currently relies on tests in Kubernetes to test the out-of-tree sidecar containers. Testing the "canary" sidecar containers is done manually by developers. There is an on-going effort to unify all sidecar containers in one repo (https://docs.google.com/document/d/1xC0Cm11WTebT9m_9yZbfzud24GJRxlp55gdR5QaMHmM/edit), which then could also import and run the E2E testing in TravisCI.

Other users are CSI driver developers.

For a full description of this proposal including an implementation plan, see this Google doc: https://docs.google.com/document/d/1lPG264ZlwPbIuFeP9P0-c0vXdmCC7wzwaapf53BbXBA/edit?usp=sharing

### PR DESCRIPTION
**Pull request by pohly**
**What this PR does / why we need it**:

There is interest in using and extending the Kubernetes test/e2e/framework also outside of Kubernetes. More and more people are developing extensions like CSI drivers that get developed and published independently from Kubernetes. Those components need a way to write and run E2E tests on a Kubernetes cluster. Trying to do that with the framework as it stands now is problematic because of the huge dependency chain.

After refactoring, other projects can create and run a custom Ginkgo suite where they:
* import test/e2e/framework as a vendor dependency without pulling in support for cloud providers and any tests (because of the huge dependency chains that a complete import causes)
* import support for cloud providers, tests and certain features separately if desired
* add custom tests without modifying the framework

**Which issue(s) this PR fixes**:
Fixes #66649

**Release note**:

```NONE
```

### CODE REVIEW

File: test/e2e/e2e.go
Comment (timothysc) [inline test/e2e/e2e.go]: Do we really need to hard smash in provider code?  
What are we using?  and given that they are moving out of tree what do we need other then a string? 

/cc @dims

File: test/e2e/e2e_test.go
Comment (timothysc) [inline test/e2e/e2e_test.go]: Please open an issue on this, it's a thorny beast.

**Review by timothysc — CHANGES_REQUESTED** [review]
So this really needs to be broken into discreet changes 1 at a time where we can make sure we vet them independently.  Right now I can't disentagle some of your changes.  Some I agree with, while others I think are the wrong approach (e.g. - cloud provider).   If you can create a breakdown of PRs where each PR is separate I can digest them individually, but as it stands this is too much churn on a core area for me to feel comfortable.

File: test/e2e/e2e.go
Comment (dims) [reply test/e2e/e2e.go]: looks like we are only referring to aws here because some of the CI jobs run the AWS specific tests, so we'll probably fail on say on vmware specific tests (but won't know until someone tries to run them)

**Comment by pohly** [comment]
"Timothy St. Clair" <notifications@github.com> writes:
> So this really needs to be broken into discreet changes 1 at a time
> where we can make sure we vet them independently.

Did you look at the individual commits or the file difference across all
commits?

I agree that this needs to be broken down, but I think I already did
that in the individual commits. I can start submitting some changes
individually, but because the later commits depend on the earlier ones,
the initial PR will have to be merged before I can commit the next one.

File: test/e2e/e2e.go
Comment (pohly) [reply test/e2e/e2e.go]: test/e2e has traditionally contained tests that only work on certain providers. Is that really changing?

In this PR, I am only trying to untangle the code so that the the provider-specific tests are separate, optional packages. But I don't want to break the existing testing, so yes, the test/e2e.test binary still needs to import the same support code for different providers as before. For example, `_ "k8s.io/kubernetes/test/e2e/framework/providers/aws"` makes is possible to say `--provider=aws`.

File: test/e2e/e2e_test.go
Comment (pohly) [reply test/e2e/e2e_test.go]: Do we need a new one or can we reopen one of the existing ones? This TODO has been in the code for a while already and previous attempts to get this resolved have stalled:

https://github.com/kubernetes/kubernetes/issues/23987
https://github.com/kubernetes/kubernetes/issues/31270

**Comment by neolit123** [comment]
@pohly 

> I agree that this needs to be broken down, but I think I already did
that in the individual commits. I can start submitting some changes
individually, but because the later commits depend on the earlier ones,
the initial PR will have to be merged before I can commit the next one.

i think @timothysc 's idea here is to just break the work chronologically into separate PRs.
once a PR i up we can review and approve it and move forward.

the downsides to that are that it might complicated the dev process for you, as you would have to maintain multiple PRs and keep them up to date with each other.
one the plus side you can open them all up from the same branch, `/hold` them and make them depend on each other.

as an example of such a PR that depends on a bunch of previous PRs:
https://github.com/kubernetes/kubernetes/pull/66219

please, let me know if that helps.
thanks

**Comment by timothysc** [comment]
> i think @timothysc 's idea here is to just break the work chronologically into separate PRs.
once a PR i up we can review and approve it and move forward. 

Indeed!  I need to take each PR and review it as a separate piece.  The commits are useful but I might push back on a sub-section.  I'd recommend converting the commits to individual PRs that can stand on their own, and we can take the reviews in a logical order.

**Comment by pohly** [comment]
@timothysc I've split out the following pieces:
- example test removal: PR #68187 (can be merged independently)
- handling extra files: PR #69103 (can be merged independently)
- config file and arguments: PR #69105 (depends on PR #68187, minor conflict with PR #69103)

**Comment by pohly** [comment]
/hold

PR #69105 needs to be merged first.

**Comment by pohly** [comment]
/hold cancel

This is it. With settings and extra file handling merged, we can now get to the actual refactoring itself :-)

https://github.com/pohly/csi-e2e has been updated such that it pulls in 8e737994768391a093b7272d2af71fe4b5639a78cfd55450a8a48f241b3d72b8, showing which dependencies are now needed when using the e2e framework outside of Kubernetes.

**Comment by pohly** [comment]
@dims you said a while back that you might be able to help test this with different cloud providers. Now would be a good time to do that for anything that isn't covered by the automated CI setup.

**Comment by dims** [comment]
ack @pohly 

/assign

File: test/e2e/e2e.go
Comment (timothysc) [reply test/e2e/e2e.go]: So I'm pretty against the idea of bin smashing in provider specifics into the main framework. In the long run, this is antithetical to where we want to go with provider code.  

If anything we want to excise it and wrap any kube specific are with an extension label, such as [storage]

File: test/e2e/framework/provider.go
Comment (timothysc) [inline test/e2e/framework/provider.go]: So umm yeah, we need to discuss this.

**Review by timothysc — COMMENTED** [review]
So we need to chat about this.  This is not really the direction where I think we should be going.  I will try to write down my notes.  Could you please attend the next sig-testing meeting.

**Comment by pohly** [comment]
"Timothy St. Clair" <notifications@github.com> writes:
> So we need to chat about this.  This is not really the direction where
> I think we should be going.  I will try to write down my notes.  Could
> you please attend the next sig-testing meeting.

Okay. I've added E2E refactoring to the agenda.

File: test/e2e/framework/provider.go
Comment (msau42) [reply test/e2e/framework/provider.go]: SIG storage has already created a sub-framework that abstracts out volume plugin (and cloud provider) setup/teardown.  I don't think we need to add an additional interface here.  For storage, it is possible to test multiple volume plugins against a single cloud provider.

Comment (msau42) [reply test/e2e/framework/provider.go]: The subframework defines an interface for different volume plugins to implement at test/e2e/storage/drivers/base.go.  I think many of these tests can be converted to use the new framework, it just needs an owner.

Comment (pohly) [reply test/e2e/framework/provider.go]: `base.go` has the interface, `in_tree.go` the implementation. Looking at the implementation for GCE, I find:
```
func (g *gcePdDriver) CreateVolume(volType testpatterns.TestVolType) interface{} {
...
        By("creating a test gce pd volume")
        vname, err := framework.CreatePDWithRetry()
...
}
```

`framework.CreatePDWithRetry` is from `pv_util.go`. In other words, this new API just wraps the existing code. It cannot replace it. Even if the tests were rewritten to not call `CreatePDWithRetry` directly, the problem would still remain.

Either the new API must be implemented differently (without direct calls to specific implementations) or we keep it as-is and connect it with the actual implementation via the API in this PR.

**Comment by spiffxp** [comment]
We discussed this during sig-testing, it _looks_ reasonable upon first glance as a step towards abstracting out the provider code.  Do we need to schedule a breakout @timothysc ?

File: test/e2e/framework/provider.go
Comment (msau42) [reply test/e2e/framework/provider.go]: We could reimplement each driver's method to call the cloud provider's methods directly, and then put each driver implementation into its own package.  My main point was I don't think the Provider interface is the right place to put PD/Volume methods because a cloud provider can support multiple volume types.

**Comment by timothysc** [comment]
@spiffxp @pohly - I'm sorry I could not make it I have overlapping meetings.  Could we have a breakout discussion on Friday this week.  Feel free to send out an invite.

**Comment by pohly** [comment]
"Timothy St. Clair" <notifications@github.com> writes:
> @spiffxp @pohly - I'm sorry I could not make it I have overlapping
> meetings.  Could we have a breakout discussion on Friday this week.

Yes, let's do that. I'm available.

> Feel free to send out an invite.

@spiffxp can you set that up? I don't know how to schedule meetings on
Zoom.

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

File: test/e2e/framework/provider.go
Comment (pohly) [reply test/e2e/framework/provider.go]: I agree that the "vendor" interface in this PR isn't ideal and something better could be designed and implemented - over time. Right now the interface closely resembles how the framework previously called vendor-specific code (`createPD` function became the `createPD` interface method). There's still time and opportunities to improve that interface while all affected code is in the same repo, but I'd like to keep that out of the current PR, otherwise it'll never get done.

**Comment by neolit123** [comment]
> Meeting was scheduled, words were said. We agreed to unblock this to move forward

great to hear!

**Comment by Katharine** [comment]
Unfortunately, this PR has broken at least running conformance tests using `--provider=skeleton` and `KUBERNETES_CONFORMANCE_TEST=y`.

**Comment by spiffxp** [comment]
ref: https://github.com/kubernetes/kubernetes/pull/70042 which appears to fix the above

**Pull request merged**
